builtin/credential/okta/backend.go

```diff
@@ -115,9 +128,81 @@ func (b *backend) Login(ctx context.Context, req *logical.Request, username stri
 	case "PASSWORD_WARN":
 		oktaResponse.AddWarning("Your Okta password is in warning state and needs to be changed soon.")
 
-	case "MFA_REQUIRED", "MFA_ENROLL":
+	case "MFA_ENROLL", "MFA_ENROLL_ACTIVATE":
 		if !cfg.BypassOktaMFA {
-			return nil, logical.ErrorResponse("okta mfa required for this account but mfa bypass not set in config"), nil, nil
+			if b.Logger().IsDebug() {
+				b.Logger().Debug("auth/okta: user must enroll or complete mfa enrollment", "user", username)
+			}
+			return nil, logical.ErrorResponse("okta authentication failed: you must complete MFA enrollment to continue"), nil, nil
+		}
+
+	case "MFA_REQUIRED":
+		if cfg.BypassOktaMFA {
+			result.Status = "SUCCESS"
+			break
+		}
+
+		factorAvailable := false
+
+		var selectedFactor mfaFactor
+		// only okta push is currently supported
+		for _, v := range result.Embedded.Factors {
+			if v.Type == "push" && v.Provider == "OKTA" {
+				factorAvailable = true
+				selectedFactor = v
+			}
+		}
+
+		if !factorAvailable {
+			return nil, logical.ErrorResponse("Okta Verify Push factor is required in order to perform MFA"), nil, nil
+		}
+
+		requestPath := fmt.Sprintf("authn/factors/%s/verify", selectedFactor.Id)
+		payload := map[string]interface{}{
+			"stateToken": result.StateToken,
+		}
+		verifyReq, err := client.NewRequest("POST", requestPath, payload)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+
+		rsp, err := client.Do(verifyReq, &result)
+		if err != nil {
+			return nil, logical.ErrorResponse(fmt.Sprintf("Okta auth failed: %v", err)), nil, nil
+		}
+		if rsp == nil {
+			return nil, logical.ErrorResponse("okta auth backend unexpected failure"), nil, nil
+		}
+		for result.Status == "MFA_CHALLENGE" {
+			switch result.FactorResult {
+			case "WAITING":
+				verifyReq, err := client.NewRequest("POST", requestPath, payload)
+				rsp, err := client.Do(verifyReq, &result)
+				if err != nil {
+					return nil, logical.ErrorResponse(fmt.Sprintf("Okta auth failed checking loop: %v", err)), nil, nil
+				}
+				if rsp == nil {
+					return nil, logical.ErrorResponse("okta auth backend unexpected failure"), nil, nil
+				}
+
+				select {
+				case <-time.After(500 * time.Millisecond):
+					// Continue
+				case <-ctx.Done():
+					return nil, logical.ErrorResponse("exiting pending mfa challenge"), nil, nil
+				}
+			case "REJECTED":
+				return nil, logical.ErrorResponse("multi-factor authentication denied"), nil, nil
+			case "TIMEOUT":
+				return nil, logical.ErrorResponse("failed to complete multi-factor authentication"), nil, nil
+			case "SUCCESS":
+				fallthrough
```

Comment:
fallthrough here *should* cause falling through to the default case which will return an error. I'm assuming this is not the behavior you're seeing since I assume you've successfully tested this, but, it's also unnecessary; can you just remove fallthrough here?